src/engine/controls/cuecontrol.cpp

```diff
@@ -1845,9 +1862,9 @@ void HotcueControl::slotHotcuePositionChanged(double newPosition) {
     emit hotcuePositionChanged(this, newPosition);
 }
 
-void HotcueControl::slotHotcueColorChanged(double newColorId) {
-    m_pCue->setColor(Color::kPredefinedColorsSet.predefinedColorFromId(newColorId));
-    emit hotcueColorChanged(this, newColorId);
+void HotcueControl::slotHotcueColorChanged(double newColor) {
+    m_pCue->setColor(mixxx::RgbColor(newColor));
```

Comment:
Do we need to handle the case if newColor < 0.0? Range checks? Please also add an explicit cast from double to the code type.

Reply:
Debug assertions would be the minimum to ensure that we are able to at least detect out of range values reliably.

Reply:
Aren't debug assertions for programming errors only? Controller scripts or the developer mode control dialog can set these values directly.

Reply:
Then we need a validation instead of only a debug assertion.

Reply:
This conversion from double control values to color codes should be written once and reused where needed. But it does not belong to RgbColor, control values need to be handled separately.

Reply:
Or we also add something like this with also includes are range check:
```c++
static optional_t optional(const double dblCode, optional_t defaultColor = nullopt());
```
Then I can just check like this:
```c++
mixxx::RgbColor::optional_t parsedColor = mixxx::RgbColor::optional(newColor);
if (!color) {
    return;
}
mixxx::RgbColor color = *parsedColor;
```

No idea if this is better. If not, where would you put that helper function?

Reply:
Control values may know about RgbColor. But RgbColor must not have any dependencies on control values. Even if it is only a semantic dependency on the double representation. No cyclic dependencies among unrelated components!

We can add a helper function, but this must be associated with the control value framework, i.e. located in /control/...

Reply:
Done in https://github.com/mixxxdj/mixxx/pull/2520/commits/cc2ed479b624ef113f9e8d4894eee02d1d22bdf2.